propertyestimator/workflow/decorators.py

```diff
@@ -1,136 +1,267 @@
+            (isinstance(default_value, (list, tuple, set, frozenset)) and len(default_value) <= 4)):
+
+            docstring = f'{docstring} The default value of this attribute ' \
+                        f'is ``{str(default_value)}``.'
+
+        elif default_value == UNDEFINED:
+
+            optional_string = '' if optional else ' and must be set by the user.'
+
+            docstring = f'{docstring} The default value of this attribute ' \
+                         f'is not set{optional_string}.'
+
+        if (merge_behavior == InequalityMergeBehaviour.SmallestValue or
+            merge_behavior == InequalityMergeBehaviour.LargestValue):
 
-    class ProtocolInputObject(BaseProtocolInputObject):
+            merge_docstring = ''
 
-        def __init__(self, class_attribute):
-            super().__init__(class_attribute)
+            if merge_behavior == InequalityMergeBehaviour.SmallestValue:
+                merge_docstring = 'When two protocols are merged, the smallest value of ' \
+                                        'this attribute from either protocol is retained.'
 
-            self.value_type = value_type
-            self.merge_behavior = merge_behavior
+            if merge_behavior == InequalityMergeBehaviour.SmallestValue:
+                merge_docstring = 'When two protocols are merged, the largest value of ' \
+                                  'this attribute from either protocol is retained.'
 
-    return ProtocolInputObject
+            docstring = f'{docstring} {merge_docstring}'
 
+        if optional is True:
+            docstring = f'{docstring} This input is *optional*.'
 
-def protocol_output(value_type):
-    """A custom decorator used to mark a protocol attribute as
-    an output of the protocol.
+        super().__init__(docstring, type_hint)
+
+        self.optional = optional
+        self.merge_behavior = merge_behavior
+
+        self._default_value = default_value
+
+    def __get__(self, instance, owner=None):
+
+        if instance is None:
+            # Handle the case where this is called on the class directly,
+            # rather than an instance.
+            return self
+
+        if not hasattr(instance, self._private_attribute_name):
+            setattr(instance, self._private_attribute_name, copy.deepcopy(self._default_value))
+
+        return getattr(instance, self._private_attribute_name)
+
+
+class ProtocolOutputAttribute(BaseProtocolAttribute):
+    """A descriptor used to mark an attribute of a protocol as
+    an output of that protocol.
 
     Examples
     ----------
-    To mark a property as an output:
+    To mark an attribute as an output:
 
-    >>> @protocol_output(value_type=str)
-    >>> def coordinate_file_path(self):
-    >>>     pass
+    >>> from propertyestimator.workflow.protocols import BaseProtocol
+    >>> from propertyestimator.workflow.decorators import protocol_output,
+    >>>
+    >>> class MyProtocol(BaseProtocol):
+    >>>
```

Comment:
Having a non-indented blank line here might actually break doctests/make this code non-functional